### PR TITLE
Keep position meta on & capture special variables in expanded AST

### PR DESCRIPTION
**Pull request by lukaszsamson**
`&` variables exist in the code so it makes sense to preserve the position in expansion

### CODE REVIEW

File: lib/elixir/test/elixir/kernel/expansion_test.exs [hunk 1]
```diff
@@ -1133,6 +1133,12 @@ defmodule Kernel.ExpansionTest do
       assert expand(quote(do: &unknown(&1, &2))) == {:&, [], [{:/, [], [{:unknown, [], nil}, 2]}]}
     end
 
+    test "keeps position meta on & variables" do
+      assert expand(Code.string_to_quoted!("& &1")) ==
+               {:fn, [{:line, 1}],
```
Comment (lukaszsamson) [inline lib/elixir/test/elixir/kernel/expansion_test.exs:1138]: I'm not sure meta on `fn` and `->` makes sense

File: lib/elixir/src/elixir_fn.erl [hunk 2]
```diff
@@ -128,12 +128,12 @@ validate(Meta, [{Pos, _} | _], Expected, E) ->
...
 
-escape({'&', _, [Pos]}, _E, Dict) when is_integer(Pos), Pos > 0 ->
+escape({'&', Meta, [Pos]}, _E, Dict) when is_integer(Pos), Pos > 0 ->
   % Using a nil context here to emit warnings when variable is unused.
   % This might pollute user space but is unlikely because variables
   % named :"&1" are not valid syntax.
-  Var = {list_to_atom([$& | integer_to_list(Pos)]), [], nil},
-  {Var, orddict:store(Pos, Var, Dict)};
+  Var = {list_to_atom([$& | integer_to_list(Pos)]), Meta, nil},
+  {Var, orddict:store(Pos, setelement(2, Var, []), Dict)};
```
Comment (lukaszsamson) [inline lib/elixir/src/elixir_fn.erl:136]: I didn't add meta on the left arg of `->` as that node is not present in the code

Comment (josevalim) [reply lib/elixir/src/elixir_fn.erl:136]: I would keep it on both sides, the more information, potentially the better?

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**